Add GKE quickstart guide with workarounds for version 0.7.0 (#350)
submariner-io/submariner-website: src/content/quickstart/gke/_index.md

---
date: 2020-11-13T17:55:18+01:00
title: "Google Kubernetes Engine (GCP)"
weight: 20
---

This quickstart guide covers the necessary steps to deploy two Google Kubernetes Engine (GKE) clusters on Google Cloud Platform (GCP) using `gcloud`. The clusters are named `cluster-a` and `cluster-b`. 

Once the GKE clusters are deployed, we deploy Submariner with Service Discovery to interconnect the two clusters. Note that this guide focuses on Submariner deployment on clusters with non-overlapping Pod and Service CIDRs. For connecting clusters with overlapping CIDRs, please refer to the [Submariner with Globalnet guide](./globalnet/), which was not tested for this tutorial.

{{% notice info %}}
The guide assumes that you have a working GCP account and the `gcloud` binary installed and configured. To get the binary, please visit [Google's website](https://cloud.google.com/sdk/docs/install).
{{% /notice %}}

## Cluster creation
Create two identical K8s clusters on GKE. For this demo, the following configuration was used:

> ``` bash

Reviewer: I think it's better if we can provide the whole example for just "copy&paste" for both cluster-a and cluster-b, 

Including the cidrs, etc: https://github.com/submariner-io/submariner-website/pull/33/files

so it's 'foolproof' :-)

Author: I see one problem with this: doesn't it depend on the subnets you have which IPs you can reserve?

Author: For me, Google was smart enough to generate non-conflicting CIDR ranges for both clusters

Reviewer: Hmm, I don't know, I was able to do that in the past (the other GKE guide).

Otherwise we need to grab those later from gke manually, and use them in join, .. or firewall... manual process always hurts (at least for quickstarts/demos).

Author: ok, so we keep it in. I would say that some knowledge of GKE is required to follow the guide, and if the CIDRs are the problem users will figure it out very quickly :)

gcloud container clusters create "cluster-a/b" \

>     --zone "<zone-id>" \

Reviewer: include one zone on the example, so it's just copy&paste :)

    --cluster-version "1.17.13-gke.1400" \
    --username "admin" \
    --machine-type "g1-small" \

>     --image-type "UBUNTU" \

Reviewer: nit: Is it necessary to use the ubuntu type of image?

Author: to be honest, I'm not sure. I wanted to provide the exact config I got it working with, because I don't know what could case other problems ... hm

Reviewer: :+1: I can try it later and if not necessary we can remove it on a separate PR.

    --disk-type "pd-ssd" \
    --disk-size "15" \
    --num-nodes "3" \
    --network "default" \
    --no-enable-ip-alias
```
{{% notice note %}}
Make sure to use Kubernetes version 1.17 or higher, set by `--cluster-version` (refer to the [GKE docs/ release notes](https://cloud.google.com/kubernetes-engine/docs/release-notes)). Also check for non-overlapping pod and service CIDRs between clusters. If this is not given, use the `--cluster-ipv4-cidr` [option](https://cloud.google.com/sdk/gcloud/reference/container/clusters/create#--cluster-ipv4-cidr). You can check using the command below:
{{% /notice %}}
``` bash
gcloud container clusters describe "cluster-a/b" --zone "<zone-id>" | grep -e clusterIpv4Cidr -e servicesIpv4Cidr
```

## Prepare clusters for Submariner

Next, the created clusters need some changes in order for Submariner to successfully open the IPSEC tunnel between them.

### Preparation: Node configuration

> As of version 0.7 of Submariner (the current one while writing this), Google's native CNI plugin is not directly supported. GKE clusters can be generated with Calico CNI instead, but this was not tested during this demo and therefore could hold surprises aswell.

Reviewer: ```suggestion
As of version 0.7 of Submariner (the current one while writing this), Google's native CNI plugin is not directly supported. GKE clusters can be generated with Calico CNI instead, but this was not tested during this demo and therefore could hold surprises as well.
```

Reviewer: Probably the linter will complain by the line length.

``` bash
wget https://raw.githubusercontent.com/sridhargaddam/k8sscripts/main/rp_filter_settings/update-rp-filter.sh
wget https://raw.githubusercontent.com/sridhargaddam/k8sscripts/main/rp_filter_settings/configure-rp-filter.sh
chmod +x update-rp-filter.sh
chmod +x configure-rp-filter.sh
echo KUBECONTEXT=cluster-a
./configure-rp-filter.sh
echo KUBECONTEXT=cluster-b
./configure-rp-filter.sh
```

### Preparation: Firewall configuration

Submariner requires the UDP ports 500, 4500 and 4800 to be open in both directions. Additionally the microservices' traffic needs to flow through the IPSEC tunnel as TCP packets.
Create firwewall rules on the GCP project that open the UDP ports aswell as all TCP ports with source and destination addresses originating in the participating clusters.
{{% notice note %}}
Use the command above to retrieve the `clusterIpv4Cidr` and `servicesIpv4Cidr`.
{{% /notice %}}

> ``` bash

Reviewer: Do we need to repeat those steps for both cluster-a and cluster-b? or isn't it necessary?

gcloud compute firewall-rules create "allow-tcp-in" --allow=tcp --direction=IN --source-ranges=<pod and service cidrs>

> gcloud compute firewall-rules create "allow-tcp-out" --allow=tcp --direction=OUT --destination-ranges=<pod and service cidrs>

Reviewer: If we specify the exact CIDRs during creation we can include them here directly , we can add a "#" comment explaining that those are the pod/service cidrs.

Kubecontext and the kubeconfig file can be specified (it defaults to ~/.kube/config):
``` bash
subctl deploy-broker --kubecontext <cluster-a/b> --kubeconfig <path/to/config>
```

Verify the Broker components are installed using `kubectl`:
``` bash
kubectl get crds --context <cluster-a/b> | grep -i submariner
  > clusters.submariner.io
    endpoints.submariner.io
    gateways.submariner.io
    multiclusterservices.lighthouse.submariner.io
    servicediscoveries.submariner.io
    serviceexports.lighthouse.submariner.io
    serviceimports.lighthouse.submariner.io
    submariners.submariner.io
```

``` bash
kubectl get ns --context <cluster-a/b> | grep -i submariner
  > submariner-k8s-broker
```

Now it is time to register every cluster in the future fleet to the Broker. Chose the context/ cluster you installed the Broker on and join this cluster to the Broker (the Broker could also be deployed on a cluster that does not run workloads, hence installing the Broker does not automatically add the cluster). The command tags the cluster for Submariner with an `<id>` you can chose.

> ``` bash

Reviewer: Again, I would do all the examples for copy & paste.

information with the Broker
* a `submariner-lighthouse-coredns` Deployment, which - as an external DNS server - gets forwarded requests to the `*.clusterset.
local` domain for cross-cluster communication by K8s internal DNS server

Check the DaemonSets and Deployments with the following command:
``` bash
kubectl --context <cluster-a/b> get ds,deploy -n submariner-operator
```

Repeat until every cluster is joined, in this case once. Then verify connectivity and CIDR settings within the fleet using 
``` bash
subctl show all
```
You should see all connected clusters and various information about them.

## Final workaround for KubeDNS
GKE per default uses KubeDNS for cluster-internal DNS queries. Submariner however only works with CoreDNS as of version 0.7. As a
consequence, the `*.clusterset.local` domain stub needs to be added manually to KubeDNS.
Query the ClusterIP of the `submariner-lighthouse-coredns` Service:
``` bash
kubectl --context <cluster-a/b> get svc submariner-lighthouse-coredns -n submariner-operator -o=custom-columns=ClusterIP:.spec.clusterIP
```

> Use the information to apply the following ConfigMap to both clusters:

Reviewer: ```suggestion
Use the information to apply the following ConfigMap to both clusters, replacing $CLUSTER_IP with the ClusterIP obtained for the lighthouse-coredns:
```

``` yaml
apiVersion: v1
kind: ConfigMap
data:
  stubDomains: |

>     {"clusterset.local":["<ClusterIP>"]}

Reviewer: nit: ```suggestion
    {"clusterset.local":["$CLUSTER_IP"]}
```

This is normally more obvious that it needs a substitution.

    --network "default" \
```

``` bash
gcloud container clusters create "cluster-b" \
    --zone "europe-west3-a" \
    --cluster-ipv4-cidr "10.4.0.0/14"
    --services-ipv4-cidr="10.7.240.0/20"
    --cluster-version "1.17.13-gke.1400" \
    --username "admin" \
    --machine-type "g1-small" \
    --image-type "UBUNTU" \
    --disk-type "pd-ssd" \
    --disk-size "15" \
    --num-nodes "3" \
    --network "default" \
```

{{% notice note %}}
Make sure to use Kubernetes version 1.17 or higher, set by `--cluster-version` (refer to the [GKE docs/ release notes](https://cloud.google.com/kubernetes-engine/docs/release-notes)). 
Also doublecheck for non-overlapping pod and service CIDRs between clusters using the commands below:
{{% /notice %}}

``` bash
gcloud container clusters describe "cluster-a" --zone "europe-west3-a" | grep -e clusterIpv4Cidr -e servicesIpv4Cidr
gcloud container clusters describe "cluster-b" --zone "europe-west3-a" | grep -e clusterIpv4Cidr -e servicesIpv4Cidr
```

## Prepare clusters for Submariner

Next, the created clusters need some changes in order for Submariner to successfully open the IPSEC tunnel between them.

### Preparation: Node configuration

> As of version 0.7 of Submariner (the current one while writing this), Google's native CNI plugin is not directly supported. GKE clusters can be generated with Calico CNI instead, but this was not tested during this demo and therefore could hold surprises as well.

Author: Google's native one, I did not find much info about it ...

> As of version 0.7 of Submariner (the current one while writing this), Google's native CNI plugin is not directly supported. GKE clusters can be generated with Calico CNI instead, but this was not tested during this demo and therefore could hold surprises as well.

Author: I've made it more clear that Google's native one is used.

Configuration is needed for the `eth0` interface of each node on every participating cluster. The used workaround deploys `netshoot` pods onto each node that configure the reverse path filtering. The scripts in [this](https://github.com/sridhargaddam/k8sscripts/tree/main/rp_filter_settings) Github repository need to be executed in all clusters.

``` bash
wget https://raw.githubusercontent.com/sridhargaddam/k8sscripts/main/rp_filter_settings/update-rp-filter.sh
wget https://raw.githubusercontent.com/sridhargaddam/k8sscripts/main/rp_filter_settings/configure-rp-filter.sh
chmod +x update-rp-filter.sh
chmod +x configure-rp-filter.sh
echo KUBECONTEXT=cluster-a
./configure-rp-filter.sh
echo KUBECONTEXT=cluster-b
./configure-rp-filter.sh
```

### Preparation: Firewall configuration

> Submariner requires the UDP ports 500, 4500 and 4800 to be open in both directions. Additionally the microservices' traffic needs to flow through the IPSEC tunnel as TCP packets.

Author: I'm no GKE expert, but I open the firewall rules per project and not per cluster. Maybe I have more possibilities for fine-tuning ...

> Submariner requires the UDP ports 500, 4500 and 4800 to be open in both directions. Additionally the microservices' traffic needs to flow through the IPSEC tunnel as TCP packets.

Author: https://cloud.google.com/solutions/prep-kubernetes-engine-for-prod#networks_and_subnets
Let me have a look :)

{{% notice note %}}
Use the commands above to retrieve the `clusterIpv4Cidr` and `servicesIpv4Cidr` for both clusters. If you copied the above code, you can just copy&paste the values below for the `--source-ranges`.
{{% /notice %}}

``` bash
gcloud compute firewall-rules create "allow-tcp-in" --allow=tcp --direction=IN --source-ranges=10.7.240.0/20,10.4.0.0/14,10.3.240.0/20,10.0.0.0/14
gcloud compute firewall-rules create "allow-tcp-out" --allow=tcp --direction=OUT --destination-ranges=10.7.240.0/20,10.4.0.0/14,10.3.240.0/20,10.0.0.0/14
gcloud compute firewall-rules create "udp-in-500" --allow=udp:500 --direction=IN
gcloud compute firewall-rules create "udp-in-4500" --allow=udp:4500 --direction=IN
gcloud compute firewall-rules create "udp-in-4800" --allow=udp:4800 --direction=IN
gcloud compute firewall-rules create "udp-out-500" --allow=udp:500 --direction=OUT
gcloud compute firewall-rules create "udp-out-4500" --allow=udp:4500 --direction=OUT
gcloud compute firewall-rules create "udp-out-4800" --allow=udp:4800 --direction=OUT
```

> After this, the clusters are finally ready for Submariner!

Reviewer: subctl will ask you :)

> After this, the clusters are finally ready for Submariner!

Author: Thanks! I need to add the dialogue of course! subctl will ask on the terminal which node to chose

## Deploy Submariner

{{< subctl-install >}}

Well will deploy the [Broker](https://submariner.io/architecture/broker/) on `cluster-a`.

The Broker cluster's API must be accessible by all the participating clusters, which is given by GKE standard configuration.

The command will output a file named "broker-info.subm" to the directory it is run from, which will be used to setup the IPSEC tunnel between clusters.

Kubecontext and the kubeconfig file can be specified (it defaults to ~/.kube/config):

``` bash
subctl deploy-broker --kubecontext cluster-a --kubeconfig <path/to/config>
```

Verify the Broker components are installed using `kubectl`:

``` bash
kubectl get crds --context cluster-a | grep -i submariner
  > clusters.submariner.io
    endpoints.submariner.io
    gateways.submariner.io
    multiclusterservices.lighthouse.submariner.io
    servicediscoveries.submariner.io
    serviceexports.lighthouse.submariner.io
    serviceimports.lighthouse.submariner.io
    submariners.submariner.io
```

``` bash
kubectl get ns --context cluster-a | grep -i submariner
  > submariner-k8s-broker
```

> Now it is time to register every cluster in the future fleet to the Broker.

Author: I thought fleet sounded nice, but as Submariner uses clusterset in the DNS name I agree. "cluster set" or "clusterset"?

> Now it is time to register every cluster in the future fleet to the Broker.

Reviewer: More than submariner, it's a k8s thing: https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#proposal

I had to look it up, it's clusterset :)

{{% notice note %}}
Use the commands above to retrieve the `clusterIpv4Cidr` and `servicesIpv4Cidr` for both clusters.
If you copied the above code, you can just copy&paste the values below for the `--source-ranges` and `--destination-ranges`.
{{% /notice %}}

``` bash
gcloud compute firewall-rules create "allow-tcp-in" --allow=tcp \
  --direction=IN --source-ranges=10.7.240.0/20,10.4.0.0/14,10.3.240.0/20,10.0.0.0/14

gcloud compute firewall-rules create "allow-tcp-out" --allow=tcp --direction=OUT \
  --destination-ranges=10.7.240.0/20,10.4.0.0/14,10.3.240.0/20,10.0.0.0/14

gcloud compute firewall-rules create "udp-in-500" --allow=udp:500 --direction=IN
gcloud compute firewall-rules create "udp-in-4500" --allow=udp:4500 --direction=IN
gcloud compute firewall-rules create "udp-in-4800" --allow=udp:4800 --direction=IN

gcloud compute firewall-rules create "udp-out-500" --allow=udp:500 --direction=OUT
gcloud compute firewall-rules create "udp-out-4500" --allow=udp:4500 --direction=OUT
gcloud compute firewall-rules create "udp-out-4800" --allow=udp:4800 --direction=OUT
```

After this, the clusters are finally ready for Submariner!

## Deploy Submariner

{{< subctl-install >}}

> Well will deploy the [Broker](https://submariner.io/architecture/broker/) on **cluster-a**.

Reviewer 2: ```suggestion
We will deploy the [Broker](https://submariner.io/architecture/broker/) on **cluster-a**.
```

gke-*-*    cluster-a   10.156.0.42    strongswan      10.3.240.0/20, 10.0.0.0/14    connected

NODE       HA STATUS     SUMMARY
gke-*-*    active        All connections (1) are established
```

### Final Workaround for KubeDNS

GKE uses KubeDNS by default for cluster-internal DNS queries. Submariner however only works with CoreDNS as of version 0.7. As a
consequence, the `*.clusterset.local` domain stub needs to be added manually to KubeDNS.
Query the ClusterIP of the `submariner-lighthouse-coredns` Service in **cluster-a**:

``` bash
CLUSTER_IP=$(kubectl --context cluster-a get svc submariner-lighthouse-coredns -n submariner-operator \
  -o=custom-columns=ClusterIP:.spec.clusterIP | tail -n +2)
```

Use the information to apply the following ConfigMap, automatically replacing $CLUSTER_IP with the ClusterIP obtained for
the `submariner-lighthouse-coredns`:

``` bash
cat <<EOF | kubectl --context cluster-a apply -f -
apiVersion: v1
kind: ConfigMap
data:
  stubDomains: |
    {"clusterset.local":["$CLUSTER_IP"]}
metadata:
  labels:
    addonmanager.kubernetes.io/mode: EnsureExists
  name: kube-dns
  namespace: kube-system
EOF
```

{{% notice note %}}

> Repeat the command, this time using `--context cluster-b`.

Reviewer: ```suggestion
Repeat both commands, the CLUSTER_IP discovery, and the kubectl apply this time using `--context cluster-b`. 
```
